In [35]:
import GA_func as ga
import numpy as np

In [36]:
sample = ga.generateFirstGen()

In [37]:
test_score = ga.fitness_score(pop)
print(min(test_score))

-4.315


In [38]:
# with swap mutation
pop = sample
min_value = []
score = ga.fitness_score(pop)
min_value.append(min(score))
for i in range(len(pop)):
    if (score[i] == min_value[-1]):
        break
var_min = pop[i]
count = 0
for i in range(1000):
    count += 1
    parents_index = ga.selectParent(pop, score)
    pop = ga.crossOver(pop, parents_index)
    pop = ga.mutation(pop, ga.swapMutation, 0.5)
    score = ga.fitness_score(pop)
    min_value.append(min(score))
    for i in range(len(pop)):
        if (score[i] == min_value[-1]):
            break
    var_min = pop[i]
    if abs(min_value[-1]-min_value[-2]) < 1e-20:
        break
print("Min value: ",min_value[-1]*1000)
print("With var: ",var_min)
print("After: ",count,' cycle')

Min value:  -3256.871859861017
With var:  [ 2.57710637  8.03999524 -6.6712894   7.69277132  2.98607289  9.46444228
 -0.29509262 -9.19681006 -4.51594282 -5.99818296  6.06897694 -1.5492932
 -0.37911558  8.95303322 -5.99409756 -8.98736018  3.983911    1.64224341
 -8.68925118  7.81529232 -3.32068968  0.87566674  0.46459146  4.78606277
 -5.32451756  0.40023206 -9.15861382 -9.5766856  -9.19681006  1.49121864
  4.49100325  9.35482013 -8.35735395  8.31491606 -5.71703379 -6.2639835
 -7.19803591 -9.26727569  7.84108831 -0.07445673 -3.63291797  8.58825664
  5.47369104  7.20805664 -8.70083301 -4.44256599 -8.56299583 -2.38167966
  1.92502153 -5.40594709]
After:  4  cycle


# for checking error
i = 0
for x in pop:
    if(x.shape != (50,)):
        print(i, '.', x.shape)
        i += 1

In [39]:
# with scramble mutation
pop = sample
min_value = []
score = ga.fitness_score(pop)
min_value.append(min(score))
for i in range(len(pop)):
    if (score[i] == min_value[-1]):
        break
var_min = pop[i]
count = 0
for i in range(1000):
    count += 1
    parents_index = ga.selectParent(pop, score)
    pop = ga.crossOver(pop, parents_index)
    pop = ga.mutation(pop, ga.scrambleMutation, 0.5)
    score = ga.fitness_score(pop)
    min_value.append(min(score))
    for i in range(len(pop)):
        if (score[i] == min_value[-1]):
            break
    var_min = pop[i]
    if abs(min_value[-1]-min_value[-2]) < 1e-20:
        break
print("Min value: ",min_value[-1]*1000)
print("With var: ",var_min)
print("After: ",count,' cycle')

Min value:  -197.33979234989178
With var:  [ 2.57710637 -0.12425517 -0.37911558 -0.12425517 -0.37911558 -0.37911558
 -0.37911558 -0.37911558 -0.37911558 -0.37911558 -0.29509262 -0.37911558
 -3.63291797 -0.37911558 -0.12425517 -0.37911558 -0.37911558 -0.37911558
 -0.29509262 -0.37911558 -0.37911558 -0.37911558 -0.37911558 -0.12425517
 -0.37911558 -0.29509262 -0.29509262 -0.37911558 -3.63291797 -0.37911558
 -0.29509262 -0.37911558 -0.37911558 -0.37911558 -0.37911558 -0.29509262
 -0.37911558 -0.37911558 -0.12425517 -0.37911558 -3.63291797 -0.37911558
 -0.37911558 -0.37911558 -0.37911558 -0.37911558 -3.63291797 -0.37911558
 -3.55762259 -4.49734644]
After:  159  cycle


In [40]:
# with inverse mutation
pop = sample
min_value = []
score = ga.fitness_score(pop)
min_value.append(min(score))
for i in range(len(pop)):
    if (score[i] == min_value[-1]):
        break
var_min = pop[i]
count = 0
for i in range(1000):
    count += 1
    parents_index = ga.selectParent(pop, score)
    pop = ga.crossOver(pop, parents_index)
    pop = ga.mutation(pop, ga.inverseMutation, 0.5)
    score = ga.fitness_score(pop)
    min_value.append(min(score))
    for i in range(len(pop)):
        if (score[i] == min_value[-1]):
            break
    var_min = pop[i]
    if abs(min_value[-1]-min_value[-2]) < 1e-20:
        break
print("Min value: ",min_value[-1]*1000)
print("With var: ",var_min)
print("After: ",count,' cycle')

Min value:  -1929.4644867741922
With var:  [ 2.57710637 -5.71703379 -8.96285004 -2.42634385 -2.42634385 -1.5492932
  6.62586239  1.12151832 -1.5492932  -1.5492932   6.62586239 -2.42634385
  1.12151832  6.62586239 -1.5492932   1.12151832 -1.5492932   1.12151832
 -1.5492932  -1.5492932   6.62586239 -2.42634385 -8.96285004  9.22061432
  9.22061432  9.22061432 -8.96285004 -2.42634385 -8.96285004 -1.5492932
  1.12151832  1.12151832 -2.42634385 -1.5492932  -8.96285004  9.22061432
 -1.5492932   1.12151832  1.12151832 -2.42634385  6.62586239 -2.42634385
 -1.5492932  -1.5492932  -2.42634385 -1.5492932  -8.96285004 -2.42634385
 -1.5492932  -9.73107012]
After:  160  cycle


In [41]:
# with random resetting
pop = sample
min_value = []
score = ga.fitness_score(pop)
min_value.append(min(score))
for i in range(len(pop)):
    if (score[i] == min_value[-1]):
        break
var_min = pop[i]
count = 0
for i in range(1000):
    count += 1
    parents_index = ga.selectParent(pop, score)
    pop = ga.crossOver(pop, parents_index)
    pop = ga.mutation(pop, ga.randomResetting, 0.5)
    score = ga.fitness_score(pop)
    min_value.append(min(score))
    for i in range(len(pop)):
        if (score[i] == min_value[-1]):
            break
    var_min = pop[i]
    if abs(min_value[-1]-min_value[-2]) < 1e-20:
        break
print("Min value: ",min_value[-1]*1000)
print("With var: ",var_min)
print("After: ",count,' cycle')

Min value:  -3670.94816357086
With var:  [  8.           8.03999524   5.          -7.          -3.
  -7.          -9.8799935   -1.           7.           0.
   5.           3.           7.          -1.          -6.
   9.          -8.          -9.          -2.           1.
 -10.           9.           6.          -7.          -5.
   2.          -9.81088748   1.           1.16714957   5.90820183
   4.          -3.         -10.          -6.          -9.
   7.          -5.67199736   3.          -1.           7.
   5.88166911   2.          -3.56528863  -3.           5.
  -9.          -3.           5.          -9.          -2.        ]
After:  45  cycle


# Test attribute: Mutation rate = 0.5
# Method: TS - MPC - SM(2)/ScM/IM/RR(3)
With the same sample with different types of mutation, after testing a few times, here is my result:

SM --  ScM  --  IM  -- RR

23(4)  ; 177(3); 448(1) ; 130(2)

45(4)  ; 151(3) ; 265(1) ; 26(2)

130(4) ; 410(4) ; 11(2)  ; 326(1)

37(3)  ; 312(4) ; 6(2)   ; 278(1)

4(2)   ; 159(4) ; 160(3) ; 45(1)

# Here is my assessment:
SM and ScM ,in general, would likely to bring out worst-fit results while IM and RR do opposite. The more loop before convergence occurs, the more likely we have a good-fit result.

